common/crypto/dkg/src/bte/proof_sharing.rs

```diff
@@ -0,0 +1,617 @@
+// Copyright 2022 - Nym Technologies SA <contact@nymtech.net>
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::bte::PublicKey;
+use crate::error::DkgError;
+use crate::interpolation::polynomial::PublicCoefficients;
+use crate::utils::{deserialize_g1, deserialize_g2, deserialize_scalar, hash_to_scalar};
+use crate::{NodeIndex, Share};
+use bls12_381::{G1Projective, G2Projective, Scalar};
+use ff::Field;
+use group::GroupEncoding;
+use rand_core::RngCore;
+use std::collections::BTreeMap;
+
+// Domain tries to follow guidelines specified by:
+// https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-hash-to-curve-11#section-3.1
+const INSTANCE_DOMAIN: &[u8] =
+    b"NYM_COCONUT_NIDKG_V01_CS01_WITH_BLS12381_XMD:SHA-256_SSWU_RO_PROOF_SECRET_SHARING_INSTANCE";
+
+const CHALLENGE_DOMAIN: &[u8] =
+    b"NYM_COCONUT_NIDKG_V01_CS01_WITH_BLS12381_XMD:SHA-256_SSWU_RO_PROOF_SECRET_SHARING_CHALLENGE";
+
+// TODO: perhaps break it down into separate arguments after all
+#[cfg_attr(test, derive(Clone))]
+pub struct Instance<'a> {
+    public_keys: &'a BTreeMap<NodeIndex, PublicKey>,
+    public_coefficients: &'a PublicCoefficients,
+    combined_randomizer: &'a G1Projective,
+    combined_ciphertexts: &'a [G1Projective],
+}
+
+impl<'a> Instance<'a> {
+    pub fn new(
+        public_keys: &'a BTreeMap<NodeIndex, PublicKey>,
+        public_coefficients: &'a PublicCoefficients,
+        combined_randomizer: &'a G1Projective,
+        combined_ciphertexts: &'a [G1Projective],
+    ) -> Instance<'a> {
+        Instance {
+            public_keys,
+            public_coefficients,
+            combined_randomizer,
+            combined_ciphertexts,
+        }
+    }
+
+    fn hash_to_scalar(&self) -> Scalar {
+        let g1s = self.public_keys.len() + 1 + self.combined_ciphertexts.len();
+        let g2s = self.public_coefficients.size();
+        let mut bytes = Vec::with_capacity(g1s * 48 + g2s * 96);
+
+        for pk in self.public_keys.values() {
+            bytes.extend_from_slice(pk.0.to_bytes().as_ref())
+        }
+        for coeff in self.public_coefficients.inner() {
+            bytes.extend_from_slice(coeff.to_bytes().as_ref())
+        }
+        bytes.extend_from_slice(self.combined_randomizer.to_bytes().as_ref());
+
+        for ciphertext in self.combined_ciphertexts {
+            bytes.extend_from_slice(ciphertext.to_bytes().as_ref())
+        }
+
+        hash_to_scalar(&bytes, INSTANCE_DOMAIN)
+    }
+
+    fn validate(&self) -> bool {
+        if self.public_keys.is_empty() || self.public_coefficients.is_empty() {
+            return false;
+        }
+
+        if self.public_keys.len() != self.combined_ciphertexts.len() {
+            return false;
+        }
+
+        true
+    }
+}
+
+#[derive(Debug)]
+#[cfg_attr(test, derive(Clone, PartialEq))]
+pub struct ProofOfSecretSharing {
+    // TODO: ask @AP for better names for those
+    f: G1Projective,
```

Comment:
Comment regarding notation: let's try to use a consistent notation for values that in the description of a cryptographic protocol are denoted with lower case letters vs upper case letters.